core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -4020,11 +4012,381 @@ class KafkaApis(val requestChannel: RequestChannel,
+    val forgottenTopics = shareFetchRequest.forgottenTopics(topicIdNames)
+
+    val newReqMetadata : ShareFetchMetadata = new ShareFetchMetadata(Uuid.fromString(memberId), shareSessionEpoch)
+    var shareFetchContext : ShareFetchContext = null
+
+    // check if the Request is Invalid. If it is, the request is failed directly here.
+    if(isInvalidShareFetchRequest) {
+      requestHelper.sendMaybeThrottle(request, shareFetchRequest.getErrorResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, Errors.INVALID_REQUEST.exception))
+      return
+    }
+
+    try {
+      // Creating the shareFetchContext for Share Session Handling. if context creation fails, the request is failed directly here.
+      shareFetchContext = sharePartitionManager.newContext(groupId, shareFetchData, forgottenTopics, newReqMetadata)
+    } catch {
+      case e: Exception => requestHelper.sendMaybeThrottle(request, shareFetchRequest.getErrorResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, e))
+        return
+    }
+
+    val erroneousAndValidPartitionData : ErroneousAndValidPartitionData = shareFetchContext.getErroneousAndValidTopicIdPartitions
+    val topicIdPartitionSeq : mutable.Set[TopicIdPartition] = mutable.Set()
+    erroneousAndValidPartitionData.erroneous.forEach {
+      case(tp, _) => if (!topicIdPartitionSeq.contains(tp)) topicIdPartitionSeq += tp
+    }
+    erroneousAndValidPartitionData.validTopicIdPartitions.forEach {
+      case(tp, _) => if (!topicIdPartitionSeq.contains(tp)) topicIdPartitionSeq += tp
+    }
+    shareFetchData.forEach {
+      case(tp, _) => if (!topicIdPartitionSeq.contains(tp)) topicIdPartitionSeq += tp
+    }
+
+    val authorizedTopics = authHelper.filterByAuthorized(
+      request.context,
+      READ,
+      TOPIC,
+      topicIdPartitionSeq
+    )(_.topicPartition.topic)
+
+    // Variable to store the topic partition wise result of fetching.
+    var fetchResult : CompletableFuture[mutable.Map[TopicIdPartition, ShareFetchResponseData.PartitionData]] =
+      CompletableFuture.completedFuture(mutable.Map.empty)
+    // Variable to store the topic partition wise result of piggybacked acknowledgements.
+    var acknowledgeResult : CompletableFuture[mutable.Map[TopicIdPartition, ShareAcknowledgeResponseData.PartitionData]] =
+      CompletableFuture.completedFuture(mutable.Map.empty)
+
+    // Handling the Acknowledgements from the ShareFetchRequest If this check is true, we are sure that this is not an
+    // Initial ShareFetch Request, otherwise the request would have been invalid.
+    if(isAcknowledgeDataPresent) {
+      val erroneous = mutable.Map[TopicIdPartition, ShareAcknowledgeResponseData.PartitionData]()
+      val acknowledgementDataFromRequest = getAcknowledgeBatchesFromShareFetchRequest(request.body[ShareFetchRequest], topicIdNames, erroneous)
+      acknowledgeResult = handleAcknowledgements(
+        acknowledgementDataFromRequest,
+        erroneous,
```

Comment:
Thanks for the review. Actually, the upcoming PR for shareAcknowledgeRequest would make it clear why it has been one this way. The acknowledgement data sent to `handleAcknowledgements` is retrieved using different methods in case of a fetch request and an acknowledge request. These methods can themselves identify some erroneous topic partitions, so that is why the map is being passed on to the method